### PR TITLE
Implement Array.prototype's copyWithin method

### PR DESCRIPTION
**Pull request by szilagyiadam**
The algorithm is based on ECMA-262 v6, 22.1.3.3

JerryScript-DCO-1.0-Signed-off-by: Adam Szilagyi aszilagy@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 1]
```diff
@@ -2301,6 +2302,123 @@ ecma_builtin_array_iterators_helper (ecma_object_t *obj_p, /**< array object */
...
+    ecma_free_value (get_value);
+
+    if (ECMA_IS_VALUE_ERROR (op_value))
+    {
+      return op_value;
+    }
+
+    ecma_free_value (op_value);
+
+    forward ? start++ : start--;
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Merge these two ternary operations.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 2]
```diff
@@ -77,6 +77,7 @@ enum
   ECMA_ARRAY_PROTOTYPE_KEYS,
   ECMA_ARRAY_PROTOTYPE_SYMBOL_ITERATOR,
   ECMA_ARRAY_PROTOTYPE_FILL,
+  ECMA_ARRAY_PROTOTYPE_COPY_WITHIN,
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: All these new feature should be guarded with `ES2015_BUILTIN` macro.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 3]
```diff
@@ -2268,6 +2269,123 @@ ecma_builtin_array_prototype_object_find (ecma_value_t predicate, /**< callback
...
+ */
+static ecma_value_t
+ecma_builtin_array_prototype_object_copy_within (const ecma_value_t args[], /**< arguments list */
+                                                 ecma_length_t args_number, /**< number of arguments */
+                                                 ecma_object_t *obj_p, /**< array object */
+                                                 uint32_t len) /**< array object's length */
+{
+  if (len == 0)
+  {
+    return ecma_make_integer_value (-1);
```
Comment (dbatyai) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Why do we return -1 here? Shouldn't this method return with the array?

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 4]
```diff
@@ -2268,6 +2269,123 @@ ecma_builtin_array_prototype_object_find (ecma_value_t predicate, /**< callback
...
+
+  if (ECMA_IS_VALUE_ERROR (error))
+  {
+    return error;
+  }
+
+  target = ecma_number_is_negative (target) ? JERRY_MAX (d_len + target, 0) : JERRY_MIN (target, d_len);
+
+  ecma_number_t start = 0;
+  ecma_number_t end = d_len;
```
Comment (dbatyai) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: I would prefer if we used `uint32_t`-s, the values are already integers in the [0, UINT32_MAX] range.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 5]
```diff
@@ -2268,6 +2269,131 @@ ecma_builtin_array_prototype_object_find (ecma_value_t predicate, /**< callback
...
+  ecma_number_t target_num;
+  ecma_value_t error = ecma_op_to_integer (args[0], &target_num);
+
+  if (ECMA_IS_VALUE_ERROR (error))
+  {
+    return error;
+  }
+
+  target_num = ecma_number_is_negative (target_num) ? JERRY_MAX (len + target_num, 0) : JERRY_MIN (target_num, len);
+  uint32_t target = ecma_number_to_uint32 (target_num);
```
Comment (dbatyai) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: `target_num` is in [0, len] range, so I'd say an `(uint32_t)` cast would be enough. Same for `start` and `end`.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 6]
```diff
@@ -2268,6 +2269,131 @@ ecma_builtin_array_prototype_object_find (ecma_value_t predicate, /**< callback
...
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_array_prototype_object_copy_within (const ecma_value_t args[], /**< arguments list */
+                                                 ecma_length_t args_number, /**< number of arguments */
+                                                 ecma_object_t *obj_p, /**< array object */
+                                                 uint32_t len) /**< array object's length */
+{
+  if (len == 0 || args_number == 0)
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Are you sure that the argument toInteger conversion can be skipped whenever the length of the target object is 0? Please test it. `[]. copyWithin( { valueof() { throw 5 })`

**Pull request merged**